server/src/controllers/projects.controllers.ts

```diff
@@ -1,94 +1,194 @@
 		}
 	};
 
-	createProject = async (req: Request, res: Response): Promise<void> => {
+	createProject = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
 		const projectsRepository = getCustomRepository(ProjectsRepository);
 
 		try {
-			const { id: userId } = req.user as UserProfile;
-			const { project } = req.body;
-			project.creatorID = userId;
+			const projectData = req.body.project as Projects;
+
+			const { id: creatorId } = req.user;
+			const user = new UserProfile();
+
+			user.id = creatorId;
+
+			const project = {
+				...new Projects(),
+				...projectData,
+				creator: user,
+				lead: user,
+				users: [user],
+			};
+
+			const validationErrors = await validateProject(project);
+
+			if (validationErrors.length > 0) {
+				next(new ErrorResponse(HttpStatusCode.UNPROCESSABLE_ENTITY, projectsErrorMessages.INVALID_DATA));
+				return;
+			}
+
+			const { key } = project;
+			const isKeyAllowedResult = await this.isKeyAllowed(key, next);
+
+			if (!isKeyAllowedResult) {
+				return;
+			}
+
 			const createdProject = await projectsRepository.createOne(project);
 			res.status(201).send(createdProject);
 		} catch (err) {
-			res.status(500).send(getWebError(err, 500));
+			next(new ErrorResponse(HttpStatusCode.INTERNAL_SERVER_ERROR, err.message));
 		}
 	};
 
-	updateProject = async (req: Request, res: Response): Promise<void> => {
+	updateProject = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
 		const projectsRepository = getCustomRepository(ProjectsRepository);
 
 		try {
 			const { project } = req.body;
-			const { id } = project;
-			await this.checkForExisting(id);
-			const updatedProject = await projectsRepository.updateOne(project);
+			const { id: projectId, key } = project;
+			const validationErrors = await validateProject(project);
+
+			if (validationErrors.length > 0) {
+				next(new ErrorResponse(HttpStatusCode.UNPROCESSABLE_ENTITY, projectsErrorMessages.INVALID_DATA));
+				return;
+			}
+
+			const prevProject = await projectsRepository.getOne(projectId);
+
+			if (prevProject === undefined) {
+				next(new ErrorResponse(HttpStatusCode.NOT_FOUND, projectsErrorMessages.PROJECT_NOT_FOUND));
+				return;
+			}
+
+			const isKeyAllowedResult = await this.isKeyAllowed(key, next, projectId);
+
+			if (!isKeyAllowedResult) {
+				return;
+			}
+
+			const { id: prevProjectLeadId } = prevProject.lead;
+			const { users: prevProjectUsers } = prevProject;
+
+			const isLeadInUsers = prevProjectUsers.find((user) => user.id === prevProjectLeadId);
+
+			if (!isLeadInUsers) {
+				next(new ErrorResponse(HttpStatusCode.UNPROCESSABLE_ENTITY, projectsErrorMessages.INVALID_DATA));
+				return;
+			}
+
+			const { id: userId } = req.user;
+
+			const isForbiddenResult = this.isForbidden(userId, prevProjectLeadId, next);
+
+			if (isForbiddenResult) {
+				return;
+			}
+
+			const projectToUpdate = { ...prevProject, ...project };
```

Comment:
Looks like _repository.save()_ support partial update, so there's no need for _prevProject_, only id and updated props. Could you please check and update?

Reply:
Partial update is supported, since some fields, for example "description" may be empty